src/algorithms/ctmrg/ctmrg.jl

```diff
@@ -177,51 +176,86 @@ ctmrg_logcancel!(log, iter, η, N) = @warnv 1 logcancel!(log, iter, η, N)
 
 """
     ctmrg_expand(state, envs, alg::CTMRG{M})
+    ctmrg_expand(dirs, state, envs::CTMRGEnv)
 
 Expand the environment by absorbing a new PEPS tensor.
 There are two modes of expansion: `M = :sequential` and `M = :simultaneous`.
 The first mode expands the environment in one direction at a time, for convenience towards
 the left. The second mode expands the environment in all four directions simultaneously.
+Alternatively, one can provide directly the `dirs` in which the environment is grown.
 """
-function ctmrg_expand(state, envs::CTMRGEnv{C,T}, ::SequentialCTMRG) where {C,T}
-    Qtype = tensormaptype(spacetype(C), 3, 3, storagetype(C))
-    Q_sw = Zygote.Buffer(envs.corners, Qtype, axes(state)...)
-    Q_nw = Zygote.Buffer(envs.corners, Qtype, axes(state)...)
-
-    directions = collect(Iterators.product(axes(state)...))
-    # @fwdthreads for (r, c) in directions
-    for (r, c) in directions
-        Q_sw[r, c] = enlarge_southwest_corner((r, c), envs, state)
-        Q_nw[r, c] = enlarge_northwest_corner((r, c), envs, state)
-    end
-
-    return copy(Q_sw), copy(Q_nw)
+function ctmrg_expand(state, envs::CTMRGEnv, ::SequentialCTMRG)
+    return ctmrg_expand([4, 1], state, envs)
+end
+function ctmrg_expand(state, envs::CTMRGEnv, ::SimultaneousCTMRG)
+    return ctmrg_expand(1:4, state, envs)
 end
-function ctmrg_expand(state, envs::CTMRGEnv{C,T}, ::SimultaneousCTMRG) where {C,T}
+# function ctmrg_expand(dirs, state, envs::CTMRGEnv)  # TODO: This doesn't AD due to length(::Nothing)...
+#     drc_combinations = collect(Iterators.product(dirs, axes(state)...))
+#     return map(idx -> enlarge_corner(idx, envs, state)(idx[1]), drc_combinations)
+# end
+function ctmrg_expand(dirs, state, envs::CTMRGEnv{C,T}) where {C,T}
     Qtype = tensormaptype(spacetype(C), 3, 3, storagetype(C))
-    Q = Zygote.Buffer(Array{Qtype,3}(undef, size(envs.corners)))
-    drc_combinations = collect(Iterators.product(axes(envs.corners)...))
-    @fwdthreads for (dir, r, c) in drc_combinations
-        Q[dir, r, c] = if dir == NORTHWEST
-            enlarge_northwest_corner((r, c), envs, state)
-        elseif dir == NORTHEAST
-            enlarge_northeast_corner((r, c), envs, state)
-        elseif dir == SOUTHEAST
-            enlarge_southeast_corner((r, c), envs, state)
-        elseif dir == SOUTHWEST
-            enlarge_southwest_corner((r, c), envs, state)
-        end
+    Q = Zygote.Buffer(Array{Qtype,3}(undef, length(dirs), size(state)...))
+    dirs_enum = [(i, dir) for (i, dir) in enumerate(dirs)]
+    drc_combinations = collect(Iterators.product(dirs_enum, axes(state)...))
+    @fwdthreads for (d, r, c) in drc_combinations
+        ec = enlarge_corner((d[2], r, c), state, envs)
+        Q[d[1], r, c] = ec(d[2])  # Explicitly construct EnlargedCorner for now
     end
-
     return copy(Q)
 end
 
+"""
+    enlarge_corner((dir, r, c), state, envs)
+
+Enlarge corner by constructing the corresponding `EnlargedCorner` struct.
+"""
```

Comment:
How would you feel about just making this a constructor overload: `EnlargedCorner(state, envs, coordinates)`?

Reply:
Oh I'd definitely prefer that! Seems much cleaner